### DISCUSSION TITLE
Эффекты

### DISCUSSION THREAD

**Discussion by HuskySmile** [vvip-68/LedPanelWiFi#29] [Идеи и предложения]
Добрый вечер. 
Идея такая, возможно-ли сделать так чтоб в случайном режиме проигрывания эффектов они не повторялись пока весь список эффектов не проиграется?
Ну типо есть к примеру 50 эффектов и нужно чтоб они все проигрались без повторения в случайном порядке и так цикл за циклом

**Comment by vvip-68** [comment]
Ну написать то всё можно. Практического смысла не вижу... Затрат на писанину много, а полезность сомнительная...
Кто помнит эти 50 в каком порядке идут...  Проще список эффектов упорядочить в нужной последовательности, а потом по кольцу крутить... Примерно то же и получится.

**Reply by saippuakauppias** [reply]
Оказывается смысл есть :) Я отключил больше 20 эффектов, т.к. некоторые частично повторяли друг друга, были странные (огонь и похожие на него), плюс некоторые не понравились. 
В итоге сейчас эффектов не много, они крутятся в случайном порядке и очень часто дублируются (из 10 эффектов 2-3 раза одно и тоже).

Мне бы тоже хотелось чтоб в ленте был "честный" рандом. 
Как пример реализации вижу (напишу на словах, реализовать не уверен что выйдет – слишком много кода в проекте уже, я лет 8 не писал ничего на плюсах): задать массив размером `MAX_EFFECTS`, заполнить его включенными эффектами (отключенные записать нулями), при включении ленты – делать `shuffle` (перемешивание) элементов массива. При переключении эффектов идти по этому массиву (если текущий элемент `effect[i] == 0`, значит это отключенный эффект – просто пропускаем его и переходим сразу к следующему `i+1`). А как только `i == MAX_EFFECTS` – сбросить `i` в 1 эффект (зациклить). При проигрывании первого эффекта – снова делать `shuffle`.
Таким образом, кажется что вмешательство в код будет минимальное и для хранения этого понадобится всего MAX_EFFECTS + 1 памяти.

**Reply by vvip-68** [reply]
В одном эффекте "Карта памяти" - еще штук сорок разных эффектов. Ну или сколько там роликов подготовили. Они будут показываться очень редко. Просто откажитесь от случайного воспроизведения. Они будут крутиться по кругу, не повторяясь.

Над предложением подумаю.

**Reply by saippuakauppias** [reply]
Да, с картой памяти и эффектами, которые внутри себя несколько содержат и настраиваются внутри шестеренки через "случайный эффект" с моим предложением косяк, согласен :(

**Comment by chipm64** [comment]
Отлаживаю SD считыватель, с эффектами. Не идет пока. Подскажите какие особенности.
Картридер в системе видится...в web появляется, среди основных программных эффектов. А файлы *.out не подхватывает.
Причин может быть много. .. начиная , что проверить контакты... 
Меня интересуют те, которые связаны с форматами.. и др. неявные.
Может можно выложить один надежный проверенный файл сюда для проверки.

**Reply by vvip-68** [reply]
С форматами - папка должна называться по ширине/высоте матрицы (например для матрицы 25x12 - так и называтьcя '25x12'), лежать корне файловой системы карточки.  Файлы должны иметь расширение '.out', не иметь в названии кириллицы. Вроде всё.

**Reply by chipm64** [reply]
Да, спасибо, помогло!) Причем работают файлы 16x16.
Идем дальше... Посмотрел JINX - довольно легко получать эффекты в формате jnx. Как бы их ловко так в out конвертнуть?
Пишут про Glediator. Да, запустил - ищу где получать формат out. Кнопочек много.. надо разбираться...

**Reply by vvip-68** [reply]
https://github.com/vvip-68/LedPanelWiFi/wiki/Экспорт-эффектов-анимации-из-ПО-«Jinx!»
https://github.com/vvip-68/LedPanelWiFi/wiki/Захват-видео-программой-«Jinx!»

**Reply by chipm64** [reply]
Клас! хорошие ссылки... без инструкции конечно тяжко понять кого куда тиснуть) А так все получается. Спасибо
Поиграюсь и нужно собирать в корпус, затем можно будет добавлять эффекты. 
Должен сказать, что интересная разработка.. Молодцы. Труд приложен большой.

**Reply by chipm64** [reply]
Почему-то не удается получить следующее: последовательные эффекты включая те, что на карте. Как будто SD -это один (формально конечно, должны воспроизвестись все эффекты на карте) из списка. Ну, эффекты чтобы вообще все воспроизводились. 
Пока так: или только внутренние эффекты или только из SD

Специально оставляю предыдущий вопрос. Он решился)) как только спросил.
Не решилось другое. Только один эффект из SD используется последовательно с внутр. эффектами. Весь список SD не идет, только тот , который выбран в настройках SD, а хотелось бы , чтобы все .... 

Кажись и это решил.. Внутри бокса SD - "последовательно", а снаружи - "автосмена эффекта"
- оно выключается, если устанавливать конкретный эффект..
Удалять вопрос что ли, теперь? ))))

**Comment by chipm64** [comment]
Насколько трудозатратно сделать переключатель "музыка SD совместно с эффектами" ? Чтобы играли мп3 с какой-то папки как звуковое сопровождение к эффектам. Это же так класно было бы.

**Reply by vvip-68** [reply]
Мне это совсем не интересно. По крайней мере на данном этапе.
Есть же масса прошивок у которых к эффектам прикручены звуки. Типа звук костра к эффекту "Огонь" или "Камин"...

**Reply by chipm64** [reply]
ясно. Имелось в виду без привязки мелодия-эффект. Просто последовательное проигрывание мелодий из определенной папки
Так дык так.
А что это за масса такая? (прошивок) Мы же тут интересуемся вашей))

**Reply by vvip-68** [reply]
Ну, если с привязкой к эффекту еще можно как-то понять/обосновать, то просто проигрывание мелодий, пока работает... эээ... гирлянда? Часы настольные?... Зачем? Почему не включить какой-нибудь плеер рядом с гораздо более качественным звуком, поскольку это будет уже специализированное устройство для музыки. Зачем гирлянде на окне музыка?  Зачем настольным часам музыка, когда можно рядом просто колонку поставить?

**Reply by chipm64** [reply]
Можно я продолжу вашу логику?)) 
Зачем такое самодельное устройство, если можно полюбоваться промышленными световыми табло коих в городах всегда найдется? 
Зачем вашему устройству вообще музыкальный модуль, если есть качественные плееры?
Зачем в устройстве будильник, если есть настольные будильники и телефоны с будильниками у каждого?
Зачем вообще заниматься программированием микроконтроллеров создавая уникальные разработки, если столько готовых законченных устройств различного назначения? "терять" на это драгоценное время жизни.
Зачем мы тут это все тогда обсуждаем? Может все-таки потому, что нам это интересно и приносит удовольствие именно в таком виде и в таком решении , какое оно есть и желание что-то привнести.. как-то изменить... 
Сам процесс! это же и есть та часть жизни которую мы проживаем)))
Не так ли? 
![WiFiLedPanel](https://github.com/user-attachments/assets/f63bc9cf-ea51-4722-9664-f31fe5a931da)
осталось только светофильтр из матового стекла, со слегка расплывчатыми пикселями мягче и красивее

**Reply by vvip-68** [reply]
Не, не так.
1. Промышленные - где-то в городе, а тут свое на окне. Причем круче чем висюльки у соседей. Это если гирлянда.
Если это часы - большое преимущество - всегда точное время, взятое с интернета. Не отстают, не убегают.
2. Музыкальный плеер - исключительно для будильника. Ну и для привлечения внимания, если в бегущей строке использован макрос звука. Впрочем, даже это уже рудимент. В предыдущей версии где было управление через MQTT можно было настроить отправку команды с какого-то сервера на проигрывание строки переданного текста со звуковым сопровождением - типа "Внимание - пришел начальник". "Пожар!!!" или что-то такое. То есть на стене висели не просто часы, а оповещатель. Чисто утилитарное назначение.
3.  Будильник - не столько потому что это часы с будильником, сколько в отличие от других "промышленных" будильников - тут будильник рассвет.
4. Программирование и микроконтроллеры - это хобби. Кто-то крючком вяжет, кто-то вышивает, кто-то торты стряпает...
5. А звук с DFPlayer - такой себе по качеству. Годится только в качестве будильника. Именно специализированные колонки звучат кратно лучше. Тем более, если звук никак не синхронизирован с событиями в устройстве, а просто играет фоном.

Девайс зачетный! Лазерная резка? Вот к НГ как раз приобрел себе хоббийный лазерный станочек - хотел сделать что-то подобное. Потому и прошивку переписал, чтобы температура справа от часов показывалась - как раз для подобного типа устройств.

По просто музыке из папки - смотрите спецификацию на DFPlayer - он управляется просто кнопками. Выводите пару кнопок на корпус, подключаете их напрямую к выводам DFPlayer - вот вам команда на включение проигрывания музыки. Никакого скетча и микроконтроллера не нужно. Если кнопки подключать через резисторы разного номинала - там по-моему до 16 команд кнопками. В общем - читайте даташит на плеер.

**Reply by vvip-68** [reply]
А потом, если когда "просто музыка", следующим шагом вам захочется, чтобы эффекты "дрыгались" под музыку или вот это в качестве частотного анализатора - столбики "прыгали".... Ну прямая логика же...

**Reply by chipm64** [reply]
Да, лазерную резку 4мм фанеру заказывал. Нарисовал детальки и дал порезать
В общем мы друг друга поняли). Кнопки ставить пожалуй руки не дойдут. 
Если торкнет, подумаю как сделать чтобы из веба включить музыку из заданной папки.   для меня тут самым сложным будет именно создать в веб элементы управления, не было опыта.
А может и нет... может переключусь на другую поделку... Как получится.
